### PR TITLE
accuracy enhancement for baichuan2-13B for npu

### PR DESCRIPTION
**Pull request by McZyWu**
## Motivation

Previously, the Baichuan-2-13B model exhibited zero effective accuracy—it failed to generate meaningful outputs for any queries and repeatedly repeated certain phrases. This issue persisted across both NPU and GPU environments, with near-zero performance on the GSM8K benchmark.

The Baichuan-2-13B model relies on ALiBi (Attention with Linear Biases) for positional encoding (it has deprecated RoPE entirely and uses bias terms in attention to encode positional information). However, the original community codebase had two critical flaws:
The attention backend (including NPU's AscendBackend) lacked native support for ALiBi bias injection.
When running Baichuan-2-13B, RoPE was disabled but the precomputed ALiBi slope values were not passed to the attention module.
These issues led to severe accuracy degradation: the model achieved only 3% accuracy on the GSM8K benchmark (across both GPU and NPU environments, no matter 0shot/1shot/5shots, with NativeBackend also at 3%), failed to generate meaningful responses to queries, and repeatedly produced repetitive phrases (effectively 0 usable accuracy).

## Modifications

Implemented full ALiBi support in the AscendBackend (NPU) attention module, including proper handling of bias terms for positional encoding, ensuring precomputed Baichuan-specific ALiBi slopes are correctly passed to the attention layer (instead of being discarded).
Aligned the attention bias logic with Baichuan-2-13B's official implementation.

## Accuracy Tests

The model now generates coherent and correct responses to individual queries.

GSM8K benchmark accuracy for ascend backend saw a dramatic improvement:
0-shot/1-shot accuracy reaches over 50% (up from the original 3% across all environments).
1-shot accuracy specifically hits 56% (matching the official Baichuan-2-13B technical report, which reported 52% accuracy with 4-shot prompting).

image
image
image
image

start command:
python -m sglang.launch_server
--model-path /xxxxxxxxx/xxxxxxxxx/model path
--trust-remote-code
--mem-fraction-static 0.9
--host 127.0.0.1
--port 8700
--tp-size 1
--device npu
--attention-backend ascend
--max-running-requests 64
--disable-radix-cache
--disable-overlap-schedule
--watchdog-timeout 9000
--base-gpu-id xxxx
--disable-cuda-graph
--enable-torch-compile
--chunked-prefill-size -1 \

test method:
1. change model path in [test/registered/ascend/llm_models/test_ascend_baichuan2_13b_chat.py]
2. python test/registered/ascend/llm_models/test_ascend_baichuan2_13b_chat.py
Supported features: TP DP
Unsupported features: Graph mode, prefix tuning...

### CODE REVIEW

File: python/sglang/srt/configs/model_config.py [hunk 1]
```diff
@@ -443,6 +443,9 @@ def _derive_model_shapes(self):
                     ):
                         setattr(self.hf_text_config, "head_dim", self.head_dim)
 
+            elif "BaichuanForCausalLM" in self.hf_config.architectures:
+                self.use_alibi = True if self.hf_config.hidden_size != 4096 else False
```
Comment (shengzhaotian) [inline python/sglang/srt/configs/model_config.py]: please apply gemini-code-assist code view suggestion

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 2]
```diff
@@ -212,6 +212,11 @@ def __init__(self, model_runner: ModelRunner):
             self.q_head_dim = (
                 self.qk_rope_head_dim + model_runner.model_config.qk_nope_head_dim
             )
+        else:
+            self.use_alibi = (
+                hasattr(model_runner.model_config, "use_alibi")
+                and model_runner.model_config.use_alibi
+            )
```
Comment (shengzhaotian) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: Consider and validate AI review comments to simplify the code

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 3]
```diff
@@ -388,6 +393,132 @@ def init_forward_metadata_replay_cuda_graph(
...
         return 0
 
+    def _generate_alibi_bias(
+        self,
+        seq_len: int,
+        slopes: torch.Tensor,
+        num_heads: int,
+        device: torch.device,
+        dtype: torch.dtype = torch.bfloat16,
+    ) -> torch.Tensor:
```
Comment (shengzhaotian) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:417]: Please refactor this function to simplify the code and avoid consecutive unsqueeze (perhaps consider view...).

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 4]
```diff
@@ -388,6 +393,132 @@ def init_forward_metadata_replay_cuda_graph(
...
+        slopes: torch.Tensor,
+        num_heads: int,
+        device: torch.device,
+        is_extend: bool = True,
+        dtype: torch.dtype = torch.bfloat16,
+    ) -> torch.Tensor:
+        if not hasattr(self, "alibi_bias") or self.alibi_bias is None:
+            MAX_LEN_ALB = 5000
+            max_seq_len = max(kv_seq_len, q_seq_len)
+            max_seq_len = max(max_seq_len, MAX_LEN_ALB)
```
Comment (shengzhaotian) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: please refactor redundant code

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 5]
```diff
@@ -388,6 +393,132 @@ def init_forward_metadata_replay_cuda_graph(
...
+            self.alibi_bias = self._generate_alibi_bias(
+                max_seq_len, slopes, num_heads, device, dtype
+            )
+
+        if not hasattr(self, "super_mask") or self.super_mask is None:
+            MAX_LEN_ALB = 5000
```
Comment (shengzhaotian) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: please refactor redundant code

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 6]
```diff
@@ -388,6 +393,132 @@ def init_forward_metadata_replay_cuda_graph(
...
+            self.super_mask = super_mask
+        if is_extend:
+            return (
+                self.alibi_bias[:, :q_seq_len, :kv_seq_len]
+                + self.super_mask[:, :q_seq_len, :kv_seq_len]
+            )
+        else:
+            return self.alibi_bias[:, :q_seq_len, :kv_seq_len]
+
+    def attn_alibi(
```
Comment (shengzhaotian) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:455]: why use native attention ? The native implementation has limited performance. You are advised to use the fused operator implementation. If the operator does not support this, then consider the native implementation.

File: python/sglang/srt/models/baichuan.py [hunk 7]
```diff
@@ -84,12 +85,19 @@ def __init__(
         prefix: str = "",
     ):
         super().__init__()
-        self.gate_up_proj = MergedColumnParallelLinear(
+        self.gate_proj = ColumnParallelLinear(
```
Comment (shengzhaotian) [inline python/sglang/srt/models/baichuan.py]: Do not modify the model architecture file unless the module needs to be split !

File: python/sglang/srt/models/baichuan.py [hunk 8]
```diff
@@ -84,12 +85,19 @@ def __init__(
...
+            hidden_size,
+            intermediate_size,
+            bias=False,
+            quant_config=quant_config,
+            prefix=add_prefix("gate_proj", prefix),
+        )
+        self.up_proj = ColumnParallelLinear(
```
Comment (shengzhaotian) [inline python/sglang/srt/models/baichuan.py]: Do not modify the model architecture file unless the module needs to be split !

File: python/sglang/srt/models/baichuan.py [hunk 9]
```diff
@@ -103,13 +111,15 @@ def __init__(
                 f"Unsupported activation: {hidden_act}. "
                 "Only silu is supported for now."
             )
-        self.act_fn = SiluAndMul()
+        self.act_fn = torch.nn.functional.silu
```
Comment (shengzhaotian) [inline python/sglang/srt/models/baichuan.py]: why change this ?

File: python/sglang/srt/models/baichuan.py [hunk 10]
```diff
@@ -124,29 +134,21 @@ def __init__(
...
-        self.total_num_kv_heads = self.num_heads
-        if self.total_num_kv_heads >= tp_size:
-            # Number of KV heads is greater than TP size, so we partition
-            # the KV heads across multiple tensor parallel GPUs.
-            assert self.total_num_kv_heads % tp_size == 0
-        else:
-            # Number of KV heads is less than TP size, so we replicate
-            # the KV heads across multiple tensor parallel GPUs.
-            assert tp_size % self.total_num_kv_heads == 0
-        self.num_kv_heads = max(1, self.total_num_kv_heads // tp_size)
```
Comment (shengzhaotian) [inline python/sglang/srt/models/baichuan.py:149]: Please reconsider the modification here. When the tp size is greater than the kv head, there should be an issue with the current code.

File: python/sglang/srt/models/baichuan.py [hunk 11]
```diff
@@ -330,19 +342,21 @@ def forward(
 class BaiChuanBaseForCausalLM(nn.Module):
     packed_modules_mapping = {
         "W_pack": ["W_pack"],
-        "gate_up_proj": [
-            "gate_proj",
-            "up_proj",
-        ],
+        "gate_proj": ["gate_proj"],
```
Comment (shengzhaotian) [inline python/sglang/srt/models/baichuan.py]: Why was this optimization removed?

File: python/sglang/srt/configs/model_config.py (hunk 1 above)
Comment (McZyWu) [reply python/sglang/srt/configs/model_config.py]: Adopted. Thank you!

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 2 above)
Comment (McZyWu) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: Adopted. Thank you! like this: self.use_alibi = getattr(model_runner.model_config, "use_alibi", False)

File: python/sglang/srt/models/baichuan.py (hunk 7 above)
Comment (McZyWu) [reply python/sglang/srt/models/baichuan.py]: okay Thank you!

File: python/sglang/srt/models/baichuan.py (hunk 8 above)
Comment (McZyWu) [reply python/sglang/srt/models/baichuan.py]: okay Thank you!

File: python/sglang/srt/models/baichuan.py (hunk 9 above)
Comment (McZyWu) [reply python/sglang/srt/models/baichuan.py]: To change the method of loading weights, this activation function was altered as well, though it's not necessary.

File: python/sglang/srt/models/baichuan.py [hunk 12]
```diff
@@ -124,20 +126,19 @@ def __init__(
...
+        dtype: Optional[torch.dtype] = torch.bfloat16,
         prefix: str = "",
     ):
         super().__init__()
         self.hidden_size = hidden_size
-        tensor_model_parallel_world_size = get_tensor_model_parallel_world_size()
         tp_size = get_tensor_model_parallel_world_size()
         self.total_num_heads = num_heads
-        assert self.total_num_heads % tensor_model_parallel_world_size == 0
-        self.num_heads = self.total_num_heads // tensor_model_parallel_world_size
```
Comment (McZyWu) [inline python/sglang/srt/models/baichuan.py:135]: remove this redundant line, caz it's assigned to the value self.num_kv_heads at line 151 in the right sidebar. (mha, the same as structures in model file modeling_baichuan.py, https://modelscope.cn/models/baichuan-inc/Baichuan2-13B-Chat/file/view/master/modeling_baichuan.py?status=1)
